### PR TITLE
Document systemdeps

### LINKED ISSUE #1686: Document kiwi-systemdeps packages
## Problem description

There is no documentation about the new kiwi-systemdeps packages. We should write something up that explains what they are good for and for which use cases they are helpful

### PR DESCRIPTION
**Pull request by schaefi**
Added documentation for kiwi-systemdeps concept
    
Added information about the kiwi-systemdeps sub packages and its purpose.
This Fixes #1686

### CODE REVIEW

**Comment by schaefi** [comment]
@davidcassany I also found a small glitch in the runtime config docs and updated it

File: doc/source/concept_and_workflow/systemdeps.rst [hunk 1]
```diff
@@ -0,0 +1,96 @@
+.. _systemdeps:
+
+Host Requirements To Build Images
+---------------------------------
+
+When building OS images, several tools and sub-systems are used
+and required on the host {kiwi} is called at. For example to
```
Comment (tomschr) [inline doc/source/concept_and_workflow/systemdeps.rst:7]:
```suggestion
and required on the host {kiwi} is called at. For example, to
```

File: doc/source/concept_and_workflow/systemdeps.rst [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+build a virtual disk image tools to create a partition table
+and also tools to create filesystems needs to be present on the
+host that builds the image.
```
Comment (tomschr) [inline doc/source/concept_and_workflow/systemdeps.rst:10]: I think I know what you want to express, but the sentence looks a bit too long and complex to me. Could we simplify it?

One idea would be:

    ... to build a virtual disk images, several tools need to be available
    on the host that builds the image.
    This include tools for partition table or tools to create filesystems.

File: doc/source/concept_and_workflow/systemdeps.rst [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+is needed to build the image. Therefore a concept to help with
+the host requirements exists and is named `kiwi-systemdeps`
+
+The `kiwi-systemdeps` concept consists out of a collection of
+sub-packages provided with the `python3-kiwi` main package. Each
+individual package requires a number of tools and subsystem packages
+which belongs to the package category. There are the following
+systemdeps packages:
+
+kiwi-systemdeps-core:
```
Comment (tomschr) [inline doc/source/concept_and_workflow/systemdeps.rst:24]: As `kiwi-systemdeps-core` is a package, I would use a monospace font to indicate something "computer-related". I would apply it to every item in this list:

```suggestion
`kiwi-systemdeps-core`:
```

File: doc/source/concept_and_workflow/systemdeps.rst [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+  Installs the package managers supported on the target
+  distribution as well as the `tar` archiving tool to support
+  building the simple root archive `tbz` image type
```
Comment (tomschr) [inline doc/source/concept_and_workflow/systemdeps.rst:27]: Missing dot:

```suggestion
  building the simple root archive `tbz` image type.
```

File: doc/source/concept_and_workflow/systemdeps.rst [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+kiwi-systemdeps-containers:
+  Installs the distribution specific tool chain to build OCI
+  compliant and WSL container images. The pulled in components
+  are required for building images of the `docker` or `appx`
+  type
+
+kiwi-systemdeps-iso-media:
+  Installs all tools required or useful to build ISO filesystems.
```
Comment (tomschr) [inline doc/source/concept_and_workflow/systemdeps.rst:36]: Not sure, but I think there is something missing:

```suggestion
  Installs all tools required or are useful to build ISO filesystems.
```

File: doc/source/concept_and_workflow/systemdeps.rst [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+kiwi-systemdeps-core:
+  Installs the package managers supported on the target
+  distribution as well as the `tar` archiving tool to support
+  building the simple root archive `tbz` image type
+
+kiwi-systemdeps-containers:
+  Installs the distribution specific tool chain to build OCI
+  compliant and WSL container images. The pulled in components
+  are required for building images of the `docker` or `appx`
+  type
```
Comment (tomschr) [inline doc/source/concept_and_workflow/systemdeps.rst:33]:
```suggestion
  are required for building `docker` or `appx` images types.
```

File: doc/source/concept_and_workflow/systemdeps.rst [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+kiwi-systemdeps-iso-media:
+  Installs all tools required or useful to build ISO filesystems.
+  The package also sets a dependency to the `-core`, `-filesystems`
+  and `-bootloaders` kiwi-systemdeps packages. The pulled in
+  components are required for building images of the `iso` type
+  as well as for disk images that request the creation of an
+  install media via the `installiso` attribute.
```
Comment (tomschr) [inline doc/source/concept_and_workflow/systemdeps.rst:41]: I would try to split the long sentence into two. My suggestion:

```suggestion
  components are required for building `iso` image types.
  For disk images that request the creation of an
  install media, the `installiso` attribute is needed.
```

File: doc/source/concept_and_workflow/systemdeps.rst [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+  to the `-core` kiwi-systemdeps packages. Image types which are
+  able to boot are of type `oem` or `iso`.
+
+  .. note::
+
+     The `iso` type is an exception which might not require the
+     `-bootloaders` systemdeps. In case of the `firmware` attribute
+     to be set to `bios`, {kiwi} builds bootable ISO images still
+     based on isolinux which is provided with the `-iso-media`
+     systemdeps. However, by default any {kiwi} created ISO image
```
Comment (tomschr) [inline doc/source/concept_and_workflow/systemdeps.rst:57]:
```suggestion
     systemdeps. However, by default, any {kiwi} created ISO image
```

File: doc/source/concept_and_workflow/systemdeps.rst [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+     causes a requirement to the `-bootloaders` systemdeps.
+  
+kiwi-systemdeps-filesystems:
+  Installs all tools to create filesystems supported with {kiwi}.
+  The pulled in components are needed for any image type except
+  for those that build just archives like `docker`, `appx` or
+  `tbz`. The package also installs tools one level below the
+  actual filesystem creation toolkit. These are components to
+  manage loop devices as well as partition table setup and
+  subsystem support for e.g LVM and LUKS. The package also sets a
```
Comment (tomschr) [inline doc/source/concept_and_workflow/systemdeps.rst:68]:
```suggestion
  subsystem support like LVM and LUKS. The package also sets a
```

File: doc/source/concept_and_workflow/systemdeps.rst [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+  dependency to the `-core` kiwi-systemdeps packages.
+
+kiwi-systemdeps-disk-images:
+  Installs all tools to create virtual disks. In {kiwi} virtual disks
```
Comment (tomschr) [inline doc/source/concept_and_workflow/systemdeps.rst:72]:
```suggestion
  Installs all tools to create virtual disks. In {kiwi}, virtual disks
```

File: doc/source/concept_and_workflow/systemdeps.rst [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+  are created using the QEMU toolchain. The pulled in components are
+  required for the `oem` image type. The package also sets a dependency
+  to the `-filesystems` and `-bootloaders` kiwi-systemdeps packages.
+
+kiwi-systemdeps-image-validation:
+  Installs the `jing` tool to validate the image description. This is
+  useful for detailed error reports from {kiwi} in case of an image
+  description validation error. In addition the `anymarkup` python
```
Comment (tomschr) [inline doc/source/concept_and_workflow/systemdeps.rst:80]:
```suggestion
  description validation error. In addition, the `anymarkup` Python
```

File: doc/source/concept_and_workflow/systemdeps.rst [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+  module is installed if the the option to install recommended packages
+  is set. With `anymarkup` available, {kiwi} can also handle image
+  descriptions in another than the XML markup, e.g YAML.
```
Comment (tomschr) [inline doc/source/concept_and_workflow/systemdeps.rst:83]:
```suggestion
  descriptions in another format than the XML markup, like YAML.
```

File: doc/source/concept_and_workflow/systemdeps.rst (hunk 3 above)
Comment (tomschr) [inline doc/source/concept_and_workflow/systemdeps.rst:24]: After reading the whole list, the information therein is very similar to each item. Another, alternative presentation could be useful to make it better readable and accessible. Maybe a table or a sublist?

For example, one presentation could be a sublist:

-----
`kiwi-systemdeps-X`:

  * **Purpose**: explains for what it is used for.
  * **Installs**: what it installs
  * **Dependencies**: what dependencies it needs
 -----

I prefer list over tables as tables are a nightmare on smartphones. :smile:  Especially if you have lots of columns.

It's just an idea, don't feel tempted to just revamp all the stuff. :slightly_smiling_face:

**Review by tomschr — COMMENTED** [review]
@schaefi Hi Marcus, as usual here are my comments and suggestion. :slightly_smiling_face: Feel free to use whatever you like.

File: doc/source/concept_and_workflow/systemdeps.rst [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+host that builds the image.
+
+The number of required components depends on the selected image
+type and the *features* used with the image. We cannot expect
+the users of {kiwi} to know about each and every component that
+is needed to build the image. Therefore a concept to help with
+the host requirements exists and is named `kiwi-systemdeps`
+
+The `kiwi-systemdeps` concept consists out of a collection of
+sub-packages provided with the `python3-kiwi` main package. Each
```
Comment (davidcassany) [inline doc/source/concept_and_workflow/systemdeps.rst:19]: nit: I'd say the main package should be `python-kiwi` as this is how src package is called.

**Review by davidcassany — APPROVED** [review]
Beyond Tom's pro comments and details it looks good to me.

Regarding Tom's suggestion about making the description of subpackages shaped in a list I do not have a strong opinion. I don't expect that any potential reader of this section to read it from start to bottom, I would only read the paragraph for the package I curious about, so I do not consider it to be too verbose as I believe it is easy to read just what you need. Anyway Tom's idea of having a more schematic approach is also nice and probably slightly easier to maintain.

File: doc/source/concept_and_workflow/systemdeps.rst (hunk 3 above)
Comment (schaefi) [reply doc/source/concept_and_workflow/systemdeps.rst]: I like the idea and adapted the docs as suggested :+1:

**Comment by schaefi** [comment]
@tomschr ok all in, unless you have further feedback I would merge as soon as the tests are green

Thanks

**Pull request merged**